File: 백준/Gold/1987. 알파벳/알파벳.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	static int r, c;
	static char[][] board;
	static int answer;
	static boolean[] visited;

	static int[] dx = { 1, 0, -1, 0 };
	static int[] dy = { 0, 1, 0, -1 };

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		r = Integer.parseInt(st.nextToken());
		c = Integer.parseInt(st.nextToken());
		board = new char[r][c];
		for (int i = 0; i < r; i++) {
			String row = br.readLine();
			for (int j = 0; j < c; j++) {
				board[i][j] = row.charAt(j);
			}
		}

		answer = Integer.MIN_VALUE;
		visited = new boolean[26];
		visited[board[0][0] - 'A'] = true;
		dfs(0, 0, 1);

		System.out.println(answer);
	}

	private static void dfs(int y, int x, int cnt) {
		answer = Math.max(answer, cnt);

		for (int i = 0; i < 4; i++) {
			int ny = y + dy[i];
			int nx = x + dx[i];
			if (0 <= ny && ny < r && 0 <= nx && nx < c && !visited[board[ny][nx] - 'A']) {
				visited[board[ny][nx] - 'A'] = true;
				dfs(ny, nx, cnt + 1);
				visited[board[ny][nx] - 'A'] = false;
			}
		}
	}

}

File: 백준/Gold/1987. 알파벳/알파벳.py
import sys

input = sys.stdin.readline

dx = (1, 0, -1, 0)
dy = (0, 1, 0, -1)


def dfs(y, x):
    result = 0
    for i in range(4):
        ny = y + dy[i]
        nx = x + dx[i]
        if 0 <= ny < r and 0 <= nx < c:
            nextCharIndex = ord(board[ny][nx]) - ord("A")
            if visited[nextCharIndex]:
                continue

            visited[nextCharIndex] = True
            result = max(result, 1 + dfs(ny, nx))
            visited[nextCharIndex] = False

    return result


r, c = map(int, input().split())
board = [list(input().strip()) for _ in range(r)]

visited = [False] * 26
charIndex = ord(board[0][0]) - ord("A")
visited[charIndex] = True

answer = dfs(0, 0) + 1

print(answer)